rasa_addons/core/tracker_stores/botfront.py

```diff
@@ -38,134 +69,119 @@ def _start_sweeper(tracker_store, break_time):
 
 class BotfrontTrackerStore(TrackerStore):
     def __init__(self, domain, url, **kwargs):
-        self.front_project_id = kwargs.get("project_id")
-        self.front_url = url
+
+        self.project_id = kwargs.get("project_id")
         self.tracker_persist_time = kwargs.get("tracker_persist_time", 3600)
         self.max_events = kwargs.get("max_events", 100)
-
         self.trackers = {}
+        self.trackers_info = (
+            {}
+        )  # in this stucture we will keep the last index and the last timestamp of events in the db for a said tracker
         self.sweeper = Thread(target=_start_sweeper, args=(self, 30))
         self.sweeper.setDaemon(True)
         self.sweeper.start()
-
+        api_key = os.environ.get("API_KEY")
+        headers = [{"Authorization": api_key}] if api_key else []
+        self.graphql_endpoint = HTTPEndpoint(url, *headers)
         super(BotfrontTrackerStore, self).__init__(domain)
+        logger.debug("BotfrontTrackerStore tracker store created")
 
-    def _fetch_tracker(self, sender_id):
-        if not sender_id in self.trackers:
-            tracker = self._get(sender_id)
-            self.trackers[sender_id] = tracker
-
-        return self.trackers[sender_id]
-
-    def _url(self, sender_id, append):
-        return "{url}/project/{project_id}/conversations/{sender_id}/{append}".format(
-            url=self.front_url,
-            project_id=self.front_project_id,
-            sender_id=sender_id,
-            append=append,
-        )
-
-    def _get(self, sender_id):
-        logger.debug("{} -> GET request for tracker".format(sender_id))
-
+    def _graphql_query(self, query, params):
         try:
-            response = requests.get(self._url(sender_id, self.max_events), timeout=1)
-            if response.status_code == 200:
-                return response.json()  # Tracker object or None if none in database
-
-            message = (
-                response.json() if response.status_code == 400 else response.reason
-            )
-            logger.warning(
-                "Error: get request to botfront-api failed with status code {}: {}".format(
-                    response.status_code, message
-                )
-            )
-        except Exception as e:
-            logger.error(
-                "Error: get request to botfront-api failed with error: {}".format(e)
-            )
-
-        return (
-            UnsynchronizedTracker()
-        )  # Return an object to represent synchronization failure with database
+            response = self.graphql_endpoint(query, params)
+            if "errors" in response and response["errors"]:
+                raise urllib.error.URLError("Null response.")
+            return response["data"]
+        except urllib.error.URLError:
+            logger.debug(f"something went wrong with the query {GET_TRACKER}")
+            return None
```

Comment:
If you return None on error, then your fetch, insert and update methods can't do data.get(...), return an empty dict instead.